jepsen/src/jepsen/control/k8s.clj

```diff
@@ -0,0 +1,111 @@
+            [clojure.tools.logging :refer [info]])
+  (:import (jepsen.control Remote)))
+
+(defn exec
+  "Execute a shell command on a pod."
+  [context namespace pod-name {:keys [cmd] :as opts}]
+  (apply sh
+         "kubectl"
+         "exec"
+         (c/escape pod-name)
+         context
+         namespace
+         "--"
+         "sh"
+         "-c"
+         cmd
+         (if-let [in (:in opts)]
+           [:in in]
+           [])))
+
+(defn- path->pod
+  [pod-name path]
+  (str pod-name ":" path))
+
+(defn- unwrap-result
+  "Throws when shell returned with nonzero exit status."
+  [exc-type {:keys [exit] :as result}]
+  (if (zero? exit)
+    result
+    (throw+
+     (assoc result :type exc-type)
+     nil ; cause
+     "Command exited with non-zero status %d:\nSTDOUT:\n%s\n\nSTDERR:\n%s"
+     exit
+     (:out result)
+     (:err result))))
+
+(defn cp-to
+  "Copies files from the host to a pod filesystem."
+  [context namespace pod-name local-paths remote-path]
+  (doseq [local-path (flatten [local-paths])]
+    (->> (sh
+          "kubectl"
+          "cp"
+          (c/escape local-path)
+          (c/escape (path->pod pod-name remote-path))
+          context
+          namespace)
+         (unwrap-result ::copy-failed))))
+
+(defn cp-from
+  "Copies files from a pod filesystem to the host."
+  [context namespace pod-name remote-paths local-path]
+  (doseq [remote-path (flatten [remote-paths])]
+    (->> (sh
+          "kubectl"
+          "cp"
+          (c/escape (path->pod pod-name remote-path))
+          (c/escape local-path)
+          context
+          namespace)
+         (unwrap-result ::copy-failed))))
+
+(defn- or-parameter
+  "A helper function that encodes a parameter if present"
+  [p, v]
+  (if v (str "--" p "=" (c/escape v)) ""))
+
+(defrecord K8sRemote [context namespace]
+  Remote
+  (connect [this pod-name]
+    (assoc this
+           :context (or-parameter "context" context)
+           :namespace (or-parameter "namespace" namespace)
+           :pod-name pod-name))
+  (disconnect! [this]
+    (dissoc this :context :namespace :pod-name))
+  (execute! [this action]
+    (exec context namespace (:pod-name this) action))
+  (upload! [this local-paths remote-path _rest]
+    (cp-to context namespace (:pod-name this) local-paths remote-path))
+  (download! [this remote-paths local-path _rest]
+    (cp-from context namespace (:pod-name this) remote-paths local-path)))
+
+(def k8s
+  "A remote that does things via `kubectl exec` and `kubectl cp`."
+  (->K8sRemote nil nil))
+
+(defn list-pods
+  "A helper function to list all pods in a given context/namespace"
+  [context namespace]
+  (let [context (or-parameter "context" context)
+        namespace (or-parameter "namespace" namespace)
+        res (sh
+             "sh"
+             "-c"
+             (str "kubectl get pods " context " " namespace
+                  " | tail -n +2 "
+                  " | awk '{print $1}'"))]
+    (info (str "" res))
```

Comment:
You can write `(str res)` here instead. Or just `res`, I think? Do you want this to log normally?

Reply:
that was a debug left-over, it's gone now